### ISSUE TITLE
[FR] Dynamically loading artifacts (e.g. jpg images)  to html artifacts

### ISSUE THREAD

**Issue by thawro** [mlflow/mlflow#13552]
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

It would be really nice to have an option to log `html` file, which uses paths to other stored artifacts in order to load the data dynamically. It would allow to create a `html` with some slider, which would scroll over the whole images directory and show each image on specific slider position (feature highly needed for computer vision models analysis). It is currently possible via logging html with `plotly.imshow` (NHW3) arrays, however it requires uploading the huge `html` file to remote each time the "scroller animation" is logged. 

### Motivation

> #### What is the use case for this feature?
 logged `html` with some slider, which would scroll over the whole images directory and show each image on specific slider position (feature highly needed for computer vision models analysis)

> #### Why is this use case valuable to support for MLflow users in general?
Other tools like `wandb` has it already implemented via images logging with step

> #### Why is this use case valuable to support for your project(s) or organization?
Analysis of computer vision models

> #### Why is it currently difficult to achieve this use case?
I can't load the images to html, the paths are incorrect or some other thing happens

### Details

Lets say that we have script `create_html.sh` which creates html which implements a scroll over images directory. This scripts accepts the following positional arguments:
1. `dirpath` - directory path with images
2. `html_outpath` - path to output html file
3. `base_path` - base path to remove from filepaths
4. `artifacts_dirpath` - artifacts uri dirpath
(I thought that 3 and 4 would be needed to properly define the specific artifacts directories, but it doesnt work)

Then, having that script it should be possible to run the following mlflow commands:
```python
    outputs_dir = "mlflow/images"
    mlflow.log_artifacts(f"{outputs_dir}/", "images")
    bash_script = "create_html.sh"
    out_html_path = "mlflow/anim.html"
    relpath = "mlflow"
    artifact_uri = f"{mlflow.get_artifact_uri()}"
    process = subprocess.Popen(
        [bash_script, outputs_dir, out_html_path, relpath, artifact_uri],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    stdout, stderr = process.communicate()
    mlflow.log_artifact(out_html_path)
```

The script first logs some all artifacts from `mlflow/images` directory to remotes `images` directory, then creates `html` file that implements scrolling over those images and finally logs that `html` file to remote.

What i expected is to be able to see the images during scrolling on the mlflow UI, but only images `alt` tags are displayed and images `src` isnt loaded.

The `create_html.sh` script:

<details>

```bash
#!/bin/bash

dirpath=$1
html_outpath=$2
base_path=$3
artifacts_dirpath=$4

jpg_files=$(find "$dirpath" -type f \( -name "*.jpg" -o -name "*.png" \) | sed "s|^$base_path|\"$artifacts_dirpath|; s|$|\",|")

echo "$jpg_files"

cat <<EOF > "$html_outpath"
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Dynamic Image Viewer with Slider</title>
  <style>
    body {
      display: flex;
      flex-direction: column;
      align-items: center;
      font-family: Arial, sans-serif;
    }
    #image-container {
      width: 80%;
      max-width: 600px;
      margin-bottom: 20px;
      text-align: center;
    }
    #slider {
      width: 80%;
      max-width: 600px;
    }
    img {
      width: 100%;
      height: auto;
      border-radius: 8px;
    }
    #filename {
      font-weight: bold;
      margin-bottom: 10px;
    }
  </style>
</head>
<body>

  <h1>Image Viewer</h1>
  <div id="image-container">
    <div id="filename">Loading...</div>
    <img id="displayed-image" src="" alt="Image Viewer" style="display: none;">
  </div>

  <div id="file-stem" style="font-weight: bold; margin-bottom: 10px;">Filename Stem: </div>

  <input type="range" id="slider" min="0" value="0" oninput="updateImage()" style="display: none;">

  <script>
    const imagePaths = [
        $jpg_files
    ];
    // Set slider max based on the number of images
    document.getElementById('slider').max = imagePaths.length - 1;


    if (imagePaths.length > 0) {
        // Show the first image initially
        document.getElementById('slider').max = imagePaths.length - 1;
        document.getElementById('slider').style.display = 'block';
        document.getElementById('displayed-image').style.display = 'block';
        updateImage();
    } else {
        document.getElementById('filename').textContent = "No images found.";
    }

    // Function to update the displayed image based on the slider index
    function updateImage() {
      const slider = document.getElementById('slider');
      const image = document.getElementById('displayed-image');
      const filenameDisplay = document.getElementById('filename');
      const fileStemDisplay = document.getElementById('file-stem');
      
      // Get the current index from the slider
      const index = slider.value;

      // Update the image source and filename display
      image.src = imagePaths[index];
      filenameDisplay.textContent = \`Filename: \${imagePaths[index]}\`;
      
      // Extract the file stem and display it
      const fileStem = imagePaths[index].split('/').pop().split('.')[0]; // Get the file stem
      fileStemDisplay.textContent = \`\${fileStem}\`;
    }
  </script>

</body>
</html>

EOF
``` 

</details>

The example output `html` file when executed on directory with 3 images (one duplicate with different filename), but run in browser from local:

![mlflow_slider](https://github.com/user-attachments/assets/3ff20eff-5a30-4b27-9f51-9d38265757e5)

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by zuzg** [comment]
That would be a very useful feature! Hope it will be implemented soon 😊

**Comment by B-Step62** [comment]
@thawro While I understand the benefit of having nicer image reviewing UI (like wandb), I'm not convinced that logging custom HTML is the right solution. From users perspective, it requires an equivalent effort to re-implementing a brand new UI and the artifact is not maintainable/reusable. Also executing an arbitral HTML on UI leads to a huge security hole.

> It would allow to create a html with some slider, which would scroll over the whole images directory and show each image on specific slider position (feature highly needed for computer vision models analysis)

With this simple requirement, I think we can consider adding a built-in slider view to the current artifacts UI. cc: @daniellok-db

**Comment by MrCorsair3** [comment]
I totally agree with this! I can't wait for this feature

**Comment by daniellok-db** [comment]
I think this makes sense, maybe in the artifacts UI, if we detect that all artifacts in a subdirectory are images, we can render this sliding view. Then the user could just put all images they want to preview inside some folder and log it. Shall we open this up as a help-wanted issue?

**Comment by thawro** [comment]
Generally speaking, having an option to log some artifact (of any kind) with `step` and `name` argument would be amazing. It would allow to create custom sliders for any kind of logged artifact (`jpg`, `png`, `html`, or maybe even for the `mp4` videos in the future). Artifacts logged with `step` and `name` attributes could be placed in the `artifacts` (same as it is done now), but also allow to slide over all logged steps for each `name` used.  I think that this kind of a feature could be a game changer in terms of artifacts logging/viewing/tracking for ML/DL cases.

**Comment by csbrown** [comment]
@thawro you say that you currently have a workaround to this problem with plotly somehow?  Can you elaborate on your solution?

**Comment by thawro** [comment]
Use `px.imshow` with `xr.DataArray` as data

**Comment by cBournhonesque** [comment]
> While I understand the benefit of having nicer image reviewing UI (like wandb [media tracking](https://docs.wandb.ai/guides/track/log/media/)), I don't believe that logging custom HTML is the right solution. From users perspective, it requires an equivalent effort to re-implementing a brand new UI and the artifact is not maintainable/reusable. Also executing an arbitral HTML on UI leads to a huge security hole.

@B-Step62  I disagree; I think for a lot of cases the security issues are minor (mlflow hosted behind company proxy with trusted contributors) and it would very useful to be able to upload a custom html that can refer to other image artifacts. This would help tell a story instead of just having a bunch of images/plots
